File: OpenWA/sdk/java/src/main/java/com/rmyndharis/openwa/resources/ContactsResource.java
package com.rmyndharis.openwa.resources;

import static com.rmyndharis.openwa.http.Http.encodeSegment;

import com.rmyndharis.openwa.OpenWAClient;
import com.rmyndharis.openwa.http.HttpMethod;
import com.rmyndharis.openwa.model.CheckNumberResponse;
import com.rmyndharis.openwa.model.ContactPhoneResponse;
import com.rmyndharis.openwa.model.ContactRecord;
import com.rmyndharis.openwa.model.ListContactsQuery;
import com.rmyndharis.openwa.model.ProfilePictureResponse;
import com.rmyndharis.openwa.model.ProfilePicturesResponse;
import com.rmyndharis.openwa.model.SuccessResult;
import com.rmyndharis.openwa.model.UpsertContactRequest;
import java.util.List;
import java.util.Map;

/** Contacts resource — contact lookup and management. */
public final class ContactsResource {
    private final OpenWAClient client;

    public ContactsResource(OpenWAClient client) {
        this.client = client;
    }

    /** List contacts known to the session. */
    public List<ContactRecord> list(String sessionId, ListContactsQuery query) {
        return client.requestList(
            HttpMethod.GET,
            "/api/sessions/" + encodeSegment(sessionId) + "/contacts",
            query,
            null,
            ContactRecord.class);
    }

    /** Get details for a single contact by id (JID). */
    public ContactRecord get(String sessionId, String contactId) {
        return client.request(
            HttpMethod.GET,
            "/api/sessions/" + encodeSegment(sessionId) + "/contacts/" + encodeSegment(contactId),
            null,
            null,
            ContactRecord.class);
    }

    /** Check whether a phone number is registered on WhatsApp. */
    public CheckNumberResponse check(String sessionId, String number) {
        return client.request(
            HttpMethod.GET,
            "/api/sessions/" + encodeSegment(sessionId) + "/contacts/check/" + encodeSegment(number),
            null,
            null,
            CheckNumberResponse.class);
    }

    /** Get the contact's profile picture URL (or null). */
    public ProfilePictureResponse profilePicture(String sessionId, String contactId) {
        return client.request(
            HttpMethod.GET,
            "/api/sessions/" + encodeSegment(sessionId) + "/contacts/" + encodeSegment(contactId) + "/profile-picture",
            null,
            null,
            ProfilePictureResponse.class);
    }

    /**
     * Batch-resolve profile picture URLs for up to 50 contacts in one request.
     * Returns a map of contact id → URL (null when a lookup fails).
     */
    public ProfilePicturesResponse profilePictures(String sessionId, List<String> ids) {
        return client.request(
            HttpMethod.GET,
            "/api/sessions/" + encodeSegment(sessionId) + "/contacts/profile-pictures",
            Map.of("ids", String.join(",", ids)),
            null,
            ProfilePicturesResponse.class);
    }

    /** Resolve a contact id (e.g. a {@code @lid}) to a phone number. */
    public ContactPhoneResponse phone(String sessionId, String contactId) {
        return client.request(
            HttpMethod.GET,
            "/api/sessions/" + encodeSegment(sessionId) + "/contacts/" + encodeSegment(contactId) + "/phone",
            null,
            null,
            ContactPhoneResponse.class);
    }

    /** Block a contact. Requires an OPERATOR-level key. */
    public SuccessResult block(String sessionId, String contactId) {
        return client.request(
            HttpMethod.POST,
            "/api/sessions/" + encodeSegment(sessionId) + "/contacts/" + encodeSegment(contactId) + "/block",
            null,
            null,
            SuccessResult.class);
    }

    /** Save a contact to the addressbook, or edit an existing entry. Requires an OPERATOR key. */
    public SuccessResult upsert(String sessionId, String contactId, UpsertContactRequest body) {
        return client.request(
            HttpMethod.PUT,
            "/api/sessions/" + encodeSegment(sessionId) + "/contacts/" + encodeSegment(contactId),
            null,
            body,
            SuccessResult.class);
    }

    /** Remove a contact from the addressbook. Requires an OPERATOR key. */
    public SuccessResult delete(String sessionId, String contactId) {
        return client.request(
            HttpMethod.DELETE,
            "/api/sessions/" + encodeSegment(sessionId) + "/contacts/" + encodeSegment(contactId),
            null,
            null,
            SuccessResult.class);
    }

    /** Unblock a contact. Requires an OPERATOR-level key. */
    public SuccessResult unblock(String sessionId, String contactId) {
        return client.request(
            HttpMethod.DELETE,
            "/api/sessions/" + encodeSegment(sessionId) + "/contacts/" + encodeSegment(contactId) + "/block",
            null,
            null,
            SuccessResult.class);
    }
}

File: OpenWA/sdk/php/src/Resources/ContactsResource.php
<?php

declare(strict_types=1);

namespace OpenWA\Resources;

use OpenWA\Http\HttpExecutor;

/**
 * Contacts resource — contact lookup and management.
 *
 * Backed by src/modules/contact/contact.controller.ts.
 */
class ContactsResource
{
    private HttpExecutor $http;

    public function __construct(HttpExecutor $http)
    {
        $this->http = $http;
    }

    /**
     * @param array<string,mixed> $query
     * @return array<int,array<string,mixed>>
     */
    public function list(string $sessionId, array $query = []): array
    {
        return $this->http->request('GET', "/api/sessions/{$this->http->encodeSegment($sessionId)}/contacts", $query) ?? [];
    }

    /** @return array<string,mixed> */
    public function get(string $sessionId, string $contactId): array
    {
        return $this->http->request('GET', "/api/sessions/{$this->http->encodeSegment($sessionId)}/contacts/{$this->http->encodeSegment($contactId)}");
    }

    /** @return array<string,mixed> */
    public function check(string $sessionId, string $number): array
    {
        return $this->http->request('GET', "/api/sessions/{$this->http->encodeSegment($sessionId)}/contacts/check/{$this->http->encodeSegment($number)}");
    }

    /** @return array<string,mixed> */
    public function profilePicture(string $sessionId, string $contactId): array
    {
        return $this->http->request('GET', "/api/sessions/{$this->http->encodeSegment($sessionId)}/contacts/{$this->http->encodeSegment($contactId)}/profile-picture");
    }

    /**
     * Batch-resolve profile picture URLs for up to 50 contacts in one request.
     *
     * @param list<string> $ids
     * @return array{pictures: array<string,?string>} Map of contact id → URL (null when a lookup fails).
     */
    public function profilePictures(string $sessionId, array $ids): array
    {
        return $this->http->request(
            'GET',
            "/api/sessions/{$this->http->encodeSegment($sessionId)}/contacts/profile-pictures",
            ['ids' => implode(',', $ids)]
        );
    }

    /** @return array<string,mixed> */
    public function phone(string $sessionId, string $contactId): array
    {
        return $this->http->request('GET', "/api/sessions/{$this->http->encodeSegment($sessionId)}/contacts/{$this->http->encodeSegment($contactId)}/phone");
    }

    /** @return array<string,mixed> */
    public function block(string $sessionId, string $contactId): array
    {
        return $this->http->request('POST', "/api/sessions/{$this->http->encodeSegment($sessionId)}/contacts/{$this->http->encodeSegment($contactId)}/block");
    }

    /**
     * Save a contact to the addressbook, or edit an existing entry. Requires an OPERATOR key.
     *
     * @param array<string,mixed> $body firstName and an optional lastName.
     * @return array<string,mixed>
     */
    public function upsert(string $sessionId, string $contactId, array $body): array
    {
        return $this->http->request('PUT', "/api/sessions/{$this->http->encodeSegment($sessionId)}/contacts/{$this->http->encodeSegment($contactId)}", [], $body) ?? [];
    }

    /**
     * Remove a contact from the addressbook. Requires an OPERATOR key.
     *
     * @return array<string,mixed>
     */
    public function delete(string $sessionId, string $contactId): array
    {
        return $this->http->request('DELETE', "/api/sessions/{$this->http->encodeSegment($sessionId)}/contacts/{$this->http->encodeSegment($contactId)}") ?? [];
    }

    /** @return array<string,mixed> */
    public function unblock(string $sessionId, string $contactId): array
    {
        return $this->http->request('DELETE', "/api/sessions/{$this->http->encodeSegment($sessionId)}/contacts/{$this->http->encodeSegment($contactId)}/block");
    }
}
